text/0000-emit-crate-deps.md

```diff
@@ -0,0 +1,62 @@
+- Feature Name:
+- Start Date: 2016-05-20
+- RFC PR: (leave this empty)
+- Rust Issue: (leave this empty)
+
+# Summary
+[summary]: #summary
+
+Add a compiler flag that emits crate dependencies on a best-effort basis.
+
+# Motivation
+[motivation]: #motivation
+
+When working with projects that consist of dozens or hundreds of compilation
+units, it is useful to have some kind of automatic dependency management in the
+build system. Otherwise such information has to be maintained in at least two
+locations: Once in the compilation unit itself via an `extern crate X` statement
+and once in the build system.
+
+To this end, this RFC proposes a new compiler flag which emits the names of the
+crates a compilation unit depends on to a text file.
+
+# Detailed design
+[design]: #detailed-design
+
+Add a new compiler flag `-Z emit-crate-deps`. If the flag is passed, the
+compiler emits a file `<crate name>.crate_deps` which contains a list of all
+crates the compilation unit depends on.
+
+If the flag `-Z parse-only` is passed, the file is emitted immediately after
+parsing. Otherwise it's emitted after macro expansion.
+
+The crates are collected by walking the AST and inspecting all `extern crate`
+statements. If the statement is of the form `extern crate X as Y`, `X` but not
+`Y` is added to the list.
+
+The file contains the names of the crates separated by `\n`.
+
+# Drawbacks
+[drawbacks]: #drawbacks
+
+If `-Z parse-only` is passed, the created file is, in general, not precise. This
+is because macro expansion can itself create new `extern crate` statements.
+However, macro expansion does, in general, depend on the dependencies having
+already been compiled.
+
+The `-Z parse-only` variant is, however, essential for the first round of
+dependency resolution in the build system.
+
+I assume that the generation of `extern crate` statements is a rarely used
+features. If it is used, the problem can be mitigated by tracking only the
+generated statements separately in the build system.
```

Comment:
>I have?

Yes, steps 3 and 4 must be run in a loop since macro expansion can reveal new dependencies which can allow more macros to expand which reveal new dependencies etc.

I don't think it's worth discussing a more complicated process that allows the compiler to emit such a file once it encounters the first error at this time. I have no idea how such a process would be implemented but I know that the process I described can be implemented in about 100 lines of code. And while such simple code cannot solve the 100% case, it can solve the 99% case and allows the other 1% to be handled downstream.

It's also not productive spending more time writing comments here until those who rule over us have signaled whether they might add this feature (in one form or the other) or not. Like I said, I'm already using this feature in my driver and maintaining 100 loc every two months isn't that much work.